src/parachain/escrow.ts

```diff
@@ -148,61 +156,94 @@ export class DefaultEscrowAPI implements EscrowAPI {
             this.getStakedBalance(accountId),
             this.systemAPI.getCurrentBlockNumber()
         ]);
-        const definedAmountToLock =
-            (
-                amountToLock
-                || newMonetaryAmount(0, this.governanceCurrency as Currency<GovernanceUnit>)
-            ) as typeof userStake;
-        const newUserStake = userStake.add(definedAmountToLock);
+        // Note: the parachain uses the balance_at which combines the staked amount
+        // and staked time to calculate the share in the reward pool
+        // like `amountLocked * unlockHeight`
+        // https://github.com/interlay/interbtc/blob/1.7.3/crates/escrow/src/lib.rs#L525
+        // We need to differentiate four cases:
+        // 1. The user does not add any new stake/extend their lock time
+        // 2. The user extends only the locktime
+        // 3. The user increases only their stake
+        // 4. The user extends the locktime and increases the stake
+
+        // Case 1: no change
+        let newUserStake = userStake;
+        let newTotalStake = totalStake;
+        const newStakedBalance = stakedBalance;
+        // Case 2 - 4
+        if (amountToLock || unlockHeight) {
+            // Maybe update the staked balance and the endBlock
```

Comment:
@savudani8 when you review, I thought we could just do
```ts
newStakedBalance.amount = stakedBalance.amount.add(amountToLock);
```

But this ends up with a type error:
```
src/parachain/escrow.ts:180:64 - error TS2345: Argument of type 'MonetaryAmount<Currency<U>, U> | undefined' is not assignable to parameter of type 'MonetaryAmount<Currency<{ readonly KINT: 12; readonly Planck: 0; } | { readonly INTR: 10; readonly Planck: 0; }>, { readonly KINT: 12; readonly Planck: 0; } | { readonly INTR: 10; readonly Planck: 0; }>'.
  Type 'undefined' is not assignable to type 'MonetaryAmount<Currency<{ readonly KINT: 12; readonly Planck: 0; } | { readonly INTR: 10; readonly Planck: 0; }>, { readonly KINT: 12; readonly Planck: 0; } | { readonly INTR: 10; readonly Planck: 0; }>'.
```

It works now but it's rather lengthy/round-about way of doing this. Let me know if there is an easier way to type cast this.

Reply:
The error shows up because the conditional doesn't guarantee that `amountToLock` is defined. There is no short solution to defaulting to a zero amount. You could do:
```
const monetaryAddedStakedBalance = amountToLock || newMonetaryAmount(0, this.governanceCurrency as Currency<GovernanceUnit>);
```

Reply:
But the above might still require casting because `amountToLock` is not guaranteed to match the `governanceCurrency`. I think a better solution is to instantiate the entire API with a governanceCurrency generic.